### ISSUE TITLE
Search by squarefree conductor, discriminat and other local behaviour

### ISSUE THREAD

**Issue by jvoight** [LMFDB/lmfdb#2625]
This was a request made by a colleague when we were playing around: it would be nice if we could search for abelian varieties like we do for number fields in that we can specify good reduction away from a set of primes.

**Comment by JohnCremona** [comment]
More suggestions:

When searching for elliptic curves with certain properties it would be helpful if one could further require that the curve
1. Has squarefree conductor, or
2. Has squarefree discriminant, or
3. Has a prime with a particular Kodaira type reduction (or other local behaviour)

**Comment by roed314** [comment]
We've just added the ability to search for classical modular forms with prime/prime-power/square/squarefree level.  To support the same for elliptic curves it suffices to just add some boolean columns to the schema.  To support the original request we just need to store a set of bad primes with a GIN index.

One way to implement a search on Kodaira type reduction would to just list the reduction types as a list of strings (which we can again GIN index).  This would support looking for curves with split multiplicative reduction at some prime, but not looking for curves with split multiplicative reduction at 5.

**Comment by JohnCremona** [comment]
Right now local data for elliptic curves is stored as a list (with one entry per bad prime) of dicts (with keys 'p', 'kod', etc).  That is fine for storing and displaying but not for searching.  I have been thinking about changing this -- in effect, transposing the matrix -- to have separate columns for 'bad_primes', 'kodaira_symbols', etc -- one for each of the current keys in those dictionaries.

Would this be a good idea?  If so I would do it for both ec/Q and ec/nf.  In any case I was about to do a rewrite of the ec_nfcurves collection to add local root numbers and so this would be an opportunity to add these new columns.  (Remark: the bad primes listed are the primes dividing the disciminant of the model stored, which is a global minimal model when such a model exists, but when there is no such model there will be one prime in the list for which the reduction type is 'good'.)

**Comment by edgarcosta** [comment]
@JohnCremona, having an entry for bad_primes of the type int4[] (i.e. arrays of signed-four byte integers) will allow you to quickly search for bad/good reduction regarding a set of primes

With respect to Kodaira symbols, I suggested text[], this will allow you to search certain types of reduction.

However, we won't be able to support searches like a certain type of reduction at the prime p.

**Comment by edgarcosta** [comment]
When we closed this issue, we forgot about Vladimir Dokchitser's request to search on:

> 1. Has squarefree conductor, or
> 2. Has squarefree discriminant, or
> 3. Has a prime with a particular Kodaira type reduction (or other local behaviour)

@JohnCremona, @AndrewVSutherland  are you still open to adding the columns?

**Comment by JohnCremona** [comment]
Sure.  A true/false flag for the two square-free conditions would be very easy to add.  The last one would be much harder with the data stored as it currently is.

As I just wrote on #3238 I have no objection to anyone adding or changing columns as long as they -- at the same time -- make appropriate changes to the data upload scripts.

**Comment by edgarcosta** [comment]
Great, thanks for the quick reply.

**Comment by AndrewVSutherland** [comment]
I agree with @JohnCremona, adding booleans for square free conductor/discriminant is trivial and something we should do.  Searching on local invariants is a lot more complicated.  I'm happy to leave this as an open feature request but I don't think we should try to address this now.

**Comment by edgarcosta** [comment]
I agree with you both, and I also believe that Vlad is mostly interested in the first two items, at least that was my impression after talking with him yesterday.

**Comment by JohnCremona** [comment]
> Just to bump this, I received a query today from Michael Stoll requesting the ability to search for prime conductor curves. Perhaps someone could tackle this during an upcoming LMFDB Friday? I don't think it will take long.

Indeed:  {'num_bad_primes':1, 'semistable':True} will find these, so it's just a question of fitting this search into the grid of search options.

**Comment by JohnCremona** [comment]
Thanks for implementing this.

Perhaps we should sort the search options a bit -- it would make sens to have the "Bad p" box right under the "Conductor box", at least?

**Comment by AndrewVSutherland** [comment]
@JohnCremona How about doing this and moving rank to the top right?  I think people search on rank much more often than j-invariant.

**Comment by JohnCremona** [comment]
Sounds good -- I agree, search by j-invariant is rare except for CM where the DM discriminant is almost always more memorable!

**Comment by AndrewVSutherland** [comment]
I could also see an argument for putting semistable and potential good reduction at the top on the right (across from conductor and bad primes)

**Comment by AndrewVSutherland** [comment]
We could also just remove semistable.  And we could remove potential good reduction if we added an "integral" selector on j-invariant.  If we did this we would want to expand the knowls on conductor and j-invariant to explain that sqfree conductor means semistable and integral j-invariant means potential good reductionn.

**Comment by JohnCremona** [comment]
These suggestions are mostly just about rewording questions which have the same answer.  If there was enough room I would like to have the word "semistable" visible as an option as well as "square-free conductor", since there will be users who don't know that these are the same.  Similarly, by all means allow for search on "integral j" but have it also say "potential good reduction".

**Comment by roed314** [comment]
It would be good to update this issue to list the specific things that still need to be done, since I think many of the desired features are now available for elliptic and genus 2 curves

**Comment by AndrewVSutherland** [comment]
Two concrete tasks whose completion would let us close this PR
- ECNF should support searches with conductor norm prime/prime-power/squarefree/divides/multiple-of the way ECQ does.
- G2Q should provide these options for conductor and absolute discriminant
